### ISSUE TITLE
Array flattening for applicable source formats (e.g. AVRO, JSON)

### ISSUE THREAD

**Issue by malthe** [risingwavelabs/risingwave#6794]
### Is your feature request related to a problem? Please describe.

Some source formats (e.g. AVRO, JSON) support array as the top-level packaging. This does not directly make sense in a SQL table context, but nonetheless, we'll sometimes have such event streams and it would be a shame to require ETL tooling to process this.

### Describe the solution you'd like

A single flattening operation is a reasonable behavior.

That is, given for example a JSON input such as:
```json
[
    {"name": "bob"},
    {"name": "joe"}
]
```
We'd simply unnested this as if two separate documents were provided.

### Additional context

In Azure Blob Storage, so-called blob events are [packaged as an array with a singleton event object](https://learn.microsoft.com/en-us/azure/event-grid/event-schema-blob-storage?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&tabs=event-grid-event-schema#example-events).

Example:
```json
[{
  "topic": "/subscriptions/{subscription-id}/resourceGroups/Storage/providers/Microsoft.Storage/storageAccounts/my-storage-account",
  "subject": "/blobServices/default/containers/test-container/blobs/new-file.txt",
  "eventType": "Microsoft.Storage.BlobCreated",
  "eventTime": "2017-06-26T18:41:00.9584103Z",
  "id": "831e1650-001e-001b-66ab-eeb76e069631",
  "data": {
    "api": "PutBlockList",
    "clientRequestId": "6d79dbfb-0e37-4fc4-981f-442c9ca65760",
    "requestId": "831e1650-001e-001b-66ab-eeb76e000000",
    "eTag": "\"0x8D4BCC2E4835CD0\"",
    "contentType": "text/plain",
    "contentLength": 524288,
    "blobType": "BlockBlob",
    "url": "https://my-storage-account.blob.core.windows.net/testcontainer/new-file.txt",
    "sequencer": "00000000000004420000000000028963",
    "storageDiagnostics": {
      "batchId": "b68529f3-68cd-4744-baa4-3c0498ec19f0"
    }
  },
  "dataVersion": "",
  "metadataVersion": "1"
}]
```

**Comment by lmatz** [comment]
Thanks for sharing your use case!

I think this case can be handled after RisingWave supports operating SQL on schemaless json data in the near future.

Will update under this thread  when it can.

**Comment by fuyufjh** [comment]
> I think this case can be handled after RisingWave supports operating SQL on schemaless json data in the near future.

Or directly support it on the connector side? Using an array as the top-level packaging looks like a common practice. I think we can always unwrap it once we see an incoming entry is wrapped with array (`[]`) instead of object (`{}`) on top level.

cc. @tabVersion @neverchanje

**Comment by waruto210** [comment]
> > I think this case can be handled after RisingWave supports operating SQL on schemaless json data in the near future.
> 
> 
> 
> Or directly support it on the connector side? Using an array as the top-level packaging looks like a common practice. I think we can always unwrap it into multiple objects once we see an incoming entry wrapped with array (`[]`) instead of object (`{}`) on top level.
> 
> 
> 
> This would be much easier than extracting multiple rows from a `JSON` column.
> 
> 
> 
> cc. @tabVersion @neverchanje 

I also agree with this method.

**Comment by tabVersion** [comment]
waiting #7032

**Comment by tabVersion** [comment]
resolved in #8061
